Refactor: move to simple default builds

droidmonkey opened the pull request:
* Remove individual feature flags in favor of a single `KPXC_MINIMAL` flag that removes advanced features from the build. Basic features are no longer guarded by feature flags.
* Basic features: Auto-Type, Yubikey, KeeShare
* Advanced features include: Browser (and passkeys), SSH Agent, and Secret Service
* Networking, Documentation, and Update Checking remain as feature flags to accommodate various distro requirements.

This change also cleans up the main CMakeLists.txt by re-arranging some content and placing macros into a dedicated include file. The minimum CMake version was bumped to 3.16.0 to conform to our minimum Ubuntu support of Focal (20.04). This also allows us to default to C++20, we fall back to C++17 for Qt versions less than 5.15.0 due to lack of support.

Lastly this change removes the KEEPASSXC_BUILD_TYPE="PreRelease" which is never used. We only support "Snapshot" and "Release" now.

## Testing strategy
[NOTE]: # ( Please describe in detail how you tested your changes. )
[TIP]:  # ( We expect new code to be covered by unit tests and documented with doc blocks! )
Tested locally and with existing CI

## Type of change
[NOTE]: # ( Please remove all lines which don't apply. )
- ✅ Breaking change (causes existing functionality to change)
- ✅ Refactor (significant modification to existing code)

droidmonkey committed:
Refactor: Move to simple default builds

* Remove individual feature flags in favor of a single `KPXC_MINIMAL` flag that removes advanced features from the build. Basic features are no longer guarded by feature flags.
* Basic features: Auto-Type, Yubikey, KeeShare
* Advanced features include: Browser (and passkeys), SSH Agent, and Secret Service
* Networking, Documentation, and Update Checking remain as feature flags to accommodate various distro requirements.

This change also cleans up the main CMakeLists.txt by re-arranging some content and placing macros into a dedicated include file. The minimum CMake version was bumped to 3.16.0 to conform to our minimum Ubuntu support of Focal (20.04). This also allows us to default to C++20, we fall back to C++17 for Qt versions less than 5.15.0 due to lack of support.

Lastly this change removes the KEEPASSXC_BUILD_TYPE="PreRelease" which is never used. We only support "Snapshot" and "Release" now.

michaelk83 commented:
Why is KeeShare considered basic? It would argue that it's an advanced niche feature. It's not needed for most single-user use, and it requires network access.

I would also argue that Secret Service is rather basic for Linux users, as it's the main method for client applications to store and retrieve their credentials. Though I agree that it's less basic than Auto-Type and Yubikey support. And I agree that it can be grouped with SSH Agent and Browser integration, since they all add an IPC layer.

I would group these features like so:
- Basic (no IPC and no networking): core functionality, Yubikey, Auto-Type.
- Local IPC, no networking: above + Browser (and passkeys), SSH Agent, Secret Service.
- Full: above + networking: Update Checks, Key Share, remote DB, icon download, HIBP.

(Technically, browser integration is also rather useless without network access, but as I understand the network access isn't done by KPXC itself.)

varjolintu commented:
Browser Integration still requires including networking to the code (even if it doesn't make any connections outside, it creates local sockets), so maybe it wouldn't fit to "no networking" category.

droidmonkey commented:
`KPXC_FEATURE_NETWORK` is specifically to disable TCP/IP network traffic. So if there are guards in browser using this flag then those guards need to be removed.

droidmonkey commented:
> Why is KeeShare considered basic? It would argue that it's an advanced niche feature. It's not needed for most single-user use, and it requires network access.

It absolutely does not require network access.

KeeShare is going to be rebranded as simply "Sharing" when it gets refactored.

Don't get hung up on the Basic/Advanced wording, thought of a couple different buckets but they were all too much.

c4rlo commented on INSTALL.md:
```diff
@@ -1,19 +1,17 @@
 * make (>= 4.2) or ninja (>= 1.10)
-* g++ (>= 4.9) or clang++ (>= 6.0)
+* g++ (>= 9.3.0) or clang++ (>= 10.0)
 * asciidoctor (>= 2.0)
```
Could we also mention the minimum versions of required libraries (e.g. Qt, Botan) in here?

c4rlo commented on INSTALL.md:
````diff
@@ -55,10 +53,10 @@ To compile from source, open a **Terminal (Linux/MacOS)**, the **MSVC Tools Comm
    ```
    mkdir build
    cd build
-   cmake -DWITH_XC_ALL=ON ..
+   cmake ..
    make
````
Or:

```suggestion
   cmake -B ./build
   cmake --build ./build
```

This way it also works with alternative "generators", such as Ninja instead of Make.

c4rlo commented on INSTALL.md:
````diff
@@ -55,10 +53,10 @@ To compile from source, open a **Terminal (Linux/MacOS)**, the **MSVC Tools Comm
    ```
````
I just noticed that the point above recommends using the `latest` git tag, which however currently points to version 2.7.4...

droidmonkey replied:
You may need to pull tags on your computer, latest points to 2.7.9 on the repository. https://github.com/keepassxreboot/keepassxc/tree/latest

c4rlo replied:
Oops, indeed, sorry.

c4rlo commented on INSTALL.md:
```diff
@@ -55,10 +53,10 @@ To compile from source, open a **Terminal (Linux/MacOS)**, the **MSVC Tools Comm
+
 If you have `vcpkg` installed, add `-DCMAKE_TOOLCHAIN_FILE=${VCPKG_ROOT}/scripts/buildsystems/vcpkg.cmake` to the `cmake` command to automatically download and install all required build and runtime dependencies locally to your build directory before compiling KeePassXC. Using `vcpkg` is the preferred way to install dependencies on macOS and required on Windows if using the MSVC toolchain.
 
 For more detailed build instructions for each platform, please refer to the [GitHub wiki](https://github.com/keepassxreboot/keepassxc/wiki/Building-KeePassXC).
```
That wiki page has a lot of info duplicated. It would be nice to standardise on a single place.

c4rlo commented on CMakeLists.txt:
```diff
@@ -157,66 +123,102 @@ endif()
-    # Check for minimum Botan version
-    message(FATAL_ERROR "Botan 2.11.0 or higher is required")   
+if(UNIX)
+    check_cxx_source_compiles("#include <sys/prctl.h>
+    int main() { prctl(PR_SET_DUMPABLE, 0); return 0; }"
+            HAVE_PR_SET_DUMPABLE)
+
+    check_cxx_source_compiles("#include <malloc.h>
+    int main() { return 0; }"
+            HAVE_MALLOC_H)
+
+    check_cxx_source_compiles("#include <malloc.h>
+    int main() { malloc_usable_size(NULL); return 0; }"
+            HAVE_MALLOC_USABLE_SIZE)
+
+    check_cxx_source_compiles("#include <sys/resource.h>
+    int main() {
+      struct rlimit limit;
+      limit.rlim_cur = 0;
+      limit.rlim_max = 0;
+      setrlimit(RLIMIT_CORE, &limit);
+      return 0;
+    }" HAVE_RLIMIT_CORE)
```
Should these snippets be moved into individual files in `cmake/compiler-checks/` for consistency?

droidmonkey replied:
I considered it, but that requires a much higher cmake version to avoid the boilerplate code. I believe it is 3.20 which isn't available on most Linux distros

c4rlo commented on src/autotype/CMakeLists.txt:
```diff
@@ -1,5 +1,19 @@
+if(UNIX AND NOT APPLE AND NOT HAIKU)
+    if(WITH_X11)
         find_package(X11 REQUIRED COMPONENTS Xi XTest)
         find_package(Qt5X11Extras 5.2 REQUIRED)
```
Version spec presumably not needed since Qt 5.12 is minimum:

```suggestion
        find_package(Qt5X11Extras REQUIRED)
```

c4rlo commented on src/gui/EntryPreviewWidget.cpp:
```diff
@@ -105,10 +103,6 @@ EntryPreviewWidget::EntryPreviewWidget(QWidget* parent)
     setFocusProxy(m_ui->entryTabWidget);
-
-#if !defined(WITH_XC_KEESHARE)
-    removeTab(m_ui->groupTabWidget, m_ui->groupShareTab);
```
The `removeTab` function can now be removed.

varjolintu commented:
This branch definitely needs a rebase. I have a few compile errors with both minimal ON and OFF.

varjolintu commented:
This is needed for this branch too:
https://github.com/keepassxreboot/keepassxc/pull/11651/changes/84d1f67a841a87ff8a5f507f3fd8cdb039666f18#diff-6479042640992283633dd58fa492ee5b69173a5b7b107c0b3c4704ddd85a679bR364

phoerious reviewed (approved):
Looks fine, but the needs conflict resolution and release-tool.ps1 should be deleted.

Commit:
Fix some ifdefs

Commit:
Move .mm files back under macos/

droidmonkey committed:
Additional cleanup of CMakeLists

1. Removed always-true CMake version guard  if(CMAKE_VERSION >= 3.14.0)

2. Removed dead GCC 4.8 stack protector fallback; C++20 requires GCC 10+, and -fstack-protector-strong has been available since GCC 4.9.

3. Removed dead ASan LSan version guard for the same reason.

4. Converted set_target_properties( COMPILE_DEFINITIONS) to target_compile_definitions() on both keepassxc_core and keepassxc_gui. Both are now PRIVATE since the symbol is an internal build detail.

5. Scoped Qt5Gui_PRIVATE_INCLUDE_DIRS to keepassxc_gui preventing Qt-internal headers from leaking into every target in the build.

6. Use cmake_parse_arguments to cleanup test case definition macros. Remove test output formats as they are never used and can be controlled when tests are run.

droidmonkey commented on CMakeLists.txt:
```diff
@@ -531,14 +436,13 @@ else()
+if(Qt5Core_VERSION VERSION_LESS "5.15.0")
+    set(CMAKE_CXX_STANDARD 17)
 endif()
 
```
This will be a non-issue when we move to Qt6

droidmonkey committed:
Additional cleanup

* Applied recommendations from thirdparty reviewer

* moved /fsanitize=address to compile options instead of definitions for MSVC builds

* fix use of LAUNCHER for fdosecretsgui test

droidmonkey commented on CMakeLists.txt:
```diff
@@ -396,90 +313,69 @@ if(APPLE AND CMAKE_COMPILER_IS_CLANGXX)
+        endif()
+        add_compile_options(/permissive- /utf-8)
+        # Clang-cl does not support /MP, /Zf, or /fsanitize=address
+        if (NOT CMAKE_COMPILER_IS_CLANG_MSVC)
+            add_compile_options(/MP /Zf /fsanitize=address)
```
Per the documentation /fsanitize is fully compatible with release optimized builds: https://learn.microsoft.com/en-us/cpp/sanitizers/asan?view=msvc-170#command-prompt
